libraries/c_sdk/standard/ble/test/iot_mqtt_ble_system_test.c

```diff
@@ -0,0 +1,1040 @@
+
+/**
+ * @brief Sends an MQTT CONNECT packet over connected BLE Channel.
+ *
+ * @param[in] pContext MQTT context pointer.
+ * @param[in] pNetworkContext Network context for OpenSSL transport implementation.
+ * @param[in] createCleanSession Creates a new MQTT session if true.
+ * If false, tries to establish the existing session if there was session
+ * already present in broker.
+ */
+static void establishMqttSession( MQTTContext_t * pContext,
+                                  NetworkContext_t * pNetworkContext,
+                                  bool createCleanSession );
+
+/**
+ * @brief Handler for incoming acknowledgement packets from the broker.
+ * @param[in] pPacketInfo Info for the incoming acknowledgement packet.
+ * @param[in] packetIdentifier The ID of the incoming packet.
+ */
+static void handleAckEvents( MQTTPacketInfo_t * pPacketInfo,
+                             uint16_t packetIdentifier );
+
+/**
+ * @brief The application callback function that is expected to be invoked by the
+ * MQTT library for incoming publish and incoming acks received over the network.
+ *
+ * @param[in] pContext MQTT context pointer.
+ * @param[in] pPacketInfo Packet Info pointer for the incoming packet.
+ * @param[in] pDeserializedInfo Deserialized information from the incoming packet.
+ */
+static void eventCallback( MQTTContext_t * pContext,
+                           MQTTPacketInfo_t * pPacketInfo,
+                           MQTTDeserializedInfo_t * pDeserializedInfo );
+
+/**
+ * The function for initializing the BLE stack.
+ * The function implementation is provided by the vendor.
+ */
+extern BTStatus_t bleStackInit( void );
+
+/*
+ * Callback invoked when physical BLE connection is established.
+ */
+static void bleConnectionCallback( BTStatus_t status,
+                                   uint16_t connId,
+                                   bool connected,
+                                   BTBdaddr_t * pBa );
+
+/**
+ * Enable BLE and GATT services.
+ * Waits for the BLE phyical connection to be established.
+ */
+static void bleEnable( void );
+
+/**
+ * Disconnect and turn off BLE.
+ */
+static void bleDisable( void );
+
+/**
+ * Sets up transport interface used for sending and receiving MQTT data.
+ */
+static void setupBleTransportInterface( NetworkContext_t * pContext );
+
+/*
+ * Cleans up transport interface used for sending receiving MQTT data.
+ */
+static void teardownBleTransportInterface( NetworkContext_t * pContext );
+
+/*-----------------------------------------------------------*/
+
+static void _BLEConnectionCallback( BTStatus_t status,
+                                    uint16_t connId,
+                                    bool connected,
+                                    BTBdaddr_t * pBa )
+{
+    if( connected == true )
+    {
+        IotBle_StopAdv( NULL );
+    }
+    else
+    {
+        ( void ) IotBle_StartAdv( NULL );
+    }
+
+    isBleConnected = connected;
+    IotSemaphore_Post( &bleConnectionSem );
```

Comment:
In the new tests and library, we are moving away from the dependency on abstraction layer. Can this dependency be avoided by using the FreeRTOS native functions directly?